rfc/0001.md

```diff
@@ -0,0 +1,148 @@
+- Start Date: 2021-02-09
+- RFC PR: https://github.com/nens/threedi-modelchecker
+
+# Summary
+
+Add an API for reading from and writing to 3Di model schematisation (sqlite)
+files.
+
+# Basic examples
+
+Initialisation:
+
+```
+>>> from threedi_modelchecker import ThreediDatabase
+>>> schematisation = ThreediDatabase({"db_path": "/path/to/sqlite"})
+```
+
+Reading records: SQLAlchemy API
+```
+>>> channels = schematisation.channels
+<sqlalchemy.orm.query.Query at 0x7faa20e9fcf8>
+>>> channels.first()
+<threedi_modelchecker.threedi_model.models.Channel at 0x7faa20c88a58>
+>>> channels.all()
+[...]
+>>> channels.filter_by(display_name="577").first()
+<threedi_modelchecker.threedi_model.models.Channel at 0x7faa20c88a58>
+```
+
+Writing/updating records: SQLAlchemy Session API + context manager
+```
+>>> from threedi_modelchecker import models
+>>> with schematisation.get_session_context() as session:
+...     session.add(models.Channel(...))
+...     session.delete(schematisation.channels.first())
+```
+
+Bulk input to numpy
+```
+>>> channels.bulk_read(fields=["the_geom", "dist_calc_points"])
+{
+    "the_geom": array([pygeos.Geometry LINESTRING ..., ...]),
+    "dist_calc_points": array([100., ...]),
+}
+```
+
+# Motivation
+
+3Di has accumulated several implementations for the schematisation I/O.
+
+*Old stack*
+The code representation of the model schema lives in two places of
+https://github.com/nens/threedi. One is a django model definition used to create
+a so-called `v2_workdbs`, the second one a custom ORM that reads data from the
+sqlite during the input file generation. The django models run on a very old
+django version (1.6) that still used south migration files. It is unmaintainable.
+
+See https://github.com/nens/threedi/tree/master/lib/threedi-tools/threedi_tools
+and https://github.com/nens/threedi/tree/master/lib/threedi-spatialite-tools/threedi_spatialite_tools.
+
+*Threedi-modelchecker*
+SQLAlchemy implementation used to validate schematisations server-side.
+
+We should replace the old Inpy implementation. Because we need a 'single source
+of truth' it is a logical choice to take the new SQLAlchemy implementation.
+
+Additional requirements (outside of this RFC) are: 
+
+- Migration support: SQLAlchemy can support this through Alembic
```

Comment:
Maybe this should also be part of this RFC. The ORM always is a specific point in the migration history, and to use it you might need to migrate your DB.

Related:
1) Do we want to be able to run migrations client-side? (Would be a big plus in my view)
   If so we need a way to determine if the list of migrations are up-to-date.